src/Workspaces/Core/Portable/Workspace/Solution/Document.cs

```diff
@@ -475,17 +475,22 @@ internal Task<DocumentOptionSet> GetOptionsAsync(OptionSet solutionOptions, Canc
             //       snapshot model. once that is fixed, we can remove this workaround - https://github.com/dotnet/roslyn/issues/19284
             if (_cachedOptions == null)
             {
-                var newAsyncLazy = new AsyncLazy<DocumentOptionSet>(async c =>
-                {
-                    var optionsService = Project.Solution.Workspace.Services.GetRequiredService<IOptionService>();
-                    var documentOptionSet = await optionsService.GetUpdatedOptionSetForDocumentAsync(this, solutionOptions, c).ConfigureAwait(false);
-                    return new DocumentOptionSet(documentOptionSet, Project.Language);
-                }, cacheResult: true);
-
-                Interlocked.CompareExchange(ref _cachedOptions, newAsyncLazy, comparand: null);
+                InitializeCachedOptions(solutionOptions, cancellationToken);
             }
 
             return _cachedOptions.GetValueAsync(cancellationToken);
         }
+
+        private void InitializeCachedOptions(OptionSet solutionOptions, CancellationToken cancellationToken)
```

Comment:
Probably should add a comment that this is split into a separate function for the reason you're doing this.

Reply:
:arrow_right: Documented via an attribute which can be enforced in the future